customfuncs/javascript.go

```diff
@@ -60,32 +61,45 @@ func parseArgTypeAndValue(argDecl, argValue string) (name string, value interfac
 		}
 		return name, b, nil
 	default:
-		return "", nil, fmt.Errorf("arg_type '%s' in '<arg_name>:<arg_type>' is not supported", declParts[1])
+		return "", nil, fmt.Errorf("arg_type '%s' in '<arg_name>:<arg_type>' is not supported", argDecl[col+1:])
 	}
 }
 
+// JSProgramCache caches *goja.Program. A *goja.Program is compiled javascript and it can be used
+// across multiple goroutines and across different *goja.Runtime. If default loading cache capacity
+// is not desirable, change JSProgramCache to a loading cache with a different capacity at package
+// init time. Be mindful this will be shared across all use cases inside your process.
+var JSProgramCache *caches.LoadingCache
+
+// jsRuntimePool caches *goja.Runtime whose creation is expensive such that we want to have a pool
+// of them to amortize the initialization cost. However, a *goja.Runtime cannot be used by two/more
+// javascript's at the same time, thus the use of sync.Pool. Not user customizable.
+// var jsRuntimePool *resPool
+var jsRuntimePool sync.Pool
+
+// NodeToJSONCache caches *node.Node to JSON translations.
+var NodeToJSONCache *caches.LoadingCache
+
 // For debugging/testing purpose so we can easily disable all the caches. But not exported. We always
 // want caching in production.
-var disableCache = false
+var disableCaching = false
 
-// JSProgramCache caches *goja.Program. A *goja.Program is compiled javascript and it can be used
-// across multiple goroutines and across different *goja.Runtime.
-var JSProgramCache = caches.NewLoadingCache() // per schema so won't have too many, no need to put a hard cap.
-// JSRuntimeCache caches *goja.Runtime. A *goja.Runtime is a javascript VM. It can *not* be shared
-// across multiple goroutines.
-// TODO still not gonna work well in scenario where a service handles many requests, each of which does a simple
-// transform (that only deals with one record) involving a javascript custom_func. We had use case where each input
-// was a small json and a transform and ctx is created for each input and does a fast/simple transform. We probably
-// need to consider a workerpool for JS runtime so it can be shared globally. Consider using
-// https://github.com/gammazero/workerpool/
-var JSRuntimeCache = caches.NewLoadingCache(100) // per transform, plus expensive, a smaller cap.
-// NodeToJSONCache caches *node.Node tree to translated JSON string.
-// TODO if in the future we have *node.Node allocation recycling, then this by-addr caching won't work.
-// Ideally, we should have a node ID which refreshes upon recycling.
-var NodeToJSONCache = caches.NewLoadingCache(100) // per transform, plus expensive, a smaller cap.
+func resetCaches() {
+	JSProgramCache = caches.NewLoadingCache()
+	jsRuntimePool = sync.Pool{
+		New: func() interface{} {
+			return goja.New()
+		},
+	}
+	NodeToJSONCache = caches.NewLoadingCache()
+}
+
+func init() {
+	resetCaches()
+}
```

Comment:
What's the reason we need to involve `init()` for now? I mean didn't see it to be used.

Reply:
init is the golang built-in package initializer. Golang runtime calls it. You already forgot? 😅

Reply:
Forgot it, thanks for letting me know